### PR TITLE
fix_stuck_VC_due_to_insufficient_VC_msgs

### PR DESCRIPTION
**Pull request by HristoStaykov**
Implemented TODO for Primary to send all View Change messages it has gathered to
other replicas on Status request.
Implemented Non-primary replica to send its View Change message on request for
status from other Replicas.

### CODE REVIEW

**Comment by HristoStaykov** [comment]
After analysis of the Logs of View Change stuck problem we discovered that it is possible for non-Primary replicas to be unable to enter a New View due to lack of View Change MSGs from sufficient other Replicas (2F+2C+1). In the logs all Replicas have received the New View message 
from the new primary, but only 2 had sufficient View Change MSGs to enter the New View.
The solution implementation is as follows:
1. If a non-Primary replica is asked for status and both the receiving and sending Replicas are in the same View, but the sending replica reports that its view is not active, the receiving will send its last View Change MSG.
2. If the Primary for the New View is the receiver of the status, send all indicated as missing
   View Change messages in the senders Status message.

File: tests/apollo/test_skvbc_chaotic_startup.py [hunk 1]
```diff
@@ -303,6 +303,52 @@ async def test_f_minus_one_staggered_replicas_requesting_vc(self, bft_network, s
 
         await self._wait_for_replicas_to_generate_checkpoint(bft_network, skvbc, expected_next_primary, bft_network.all_replicas(without={initial_primary}))
 
+    @with_trio
+    @with_bft_network(start_replica_cmd_with_vc_timeout("20000"),
+                      selected_configs=lambda n, f, c: n == 7)
+    @with_constant_load
+    async def stuck_view_change_bug_recreation(self, bft_network, skvbc, nursery):
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_chaotic_startup.py:310]: if you remove the fix, does the test fail? 😄

Comment (HristoStaykov) [reply tests/apollo/test_skvbc_chaotic_startup.py:310]: yes, I also had put logs locally to verify the execution goes through the newly implemented parts after the fix :)

File: tests/apollo/test_skvbc_chaotic_startup.py [hunk 2]
```diff
@@ -303,6 +303,52 @@ async def test_f_minus_one_staggered_replicas_requesting_vc(self, bft_network, s
...
+        """
+        Test inspired by failure of a subset of replicas to enter a New View after View Change
+        due to insufficient ViewChange messages and previous no resend of ViewChange messages
+        on Status requests. To recreate the following steps are executed:
+        1) Start all 7 replicas.
+        2) Setup an adversary that blocks all incoming msgs to replicas 2, 3, 4, 5 and 6.
```
[the remark this answers is not in the document]
Comment (HristoStaykov) [reply tests/apollo/test_skvbc_chaotic_startup.py]: Definitely should be mentioned in comment, will also add comment for the waiting of View Change initiation in all replicas.

File: tests/apollo/test_skvbc_chaotic_startup.py [hunk 3]
```diff
@@ -303,6 +303,52 @@ async def test_f_minus_one_staggered_replicas_requesting_vc(self, bft_network, s
...
+
+            bft_network.start_all_replicas()
+
+            for r in bft_network.all_replicas():
+                view_of_replica = 0
+                while view_of_replica == 0:
+                    view_of_replica = await self._get_gauge(r, bft_network, 'view')
+                    await trio.sleep(seconds=0.1)
+
+        await trio.sleep(seconds=30)
```
[the remark this answers is not in the document]
Comment (HristoStaykov) [reply tests/apollo/test_skvbc_chaotic_startup.py]: It is not really necessary, while I was working on reproducing the problem I wanted to wait to see that no retransmission of View Change messages will happen, though I also checked the code. With the fix the timeout of waiting for View Change seams sufficient.

**Comment by HristoStaykov** [comment]
After discussion with Guy Gueta, I have modified the function ViewsManager::getViewChangeMsgsForView to return the the last ViewChange messages received from each replica (even if the viewIsPending(v) == false).

**Pull request merged**